python/src/rtree.rs

```diff
@@ -27,6 +29,228 @@ impl<'a> FromPyObject<'a> for RTreeMethod {
+    fn extract_bound(obj: &Bound<'py, PyAny>) -> PyResult<Self> {
+        let buffer = PyBuffer::<u8>::get_bound(obj)?;
+        if !buffer.readonly() {
+            return Err(PyValueError::new_err("Must be read-only byte buffer."));
+        }
+        if buffer.dimensions() != 1 {
+            return Err(PyValueError::new_err("Expected 1-dimensional array."));
+        }
+        // Note: this is probably superfluous for 1D array
+        if !buffer.is_c_contiguous() {
+            return Err(PyValueError::new_err("Expected c-contiguous array."));
+        }
+        if buffer.len_bytes() == 0 {
+            return Err(PyValueError::new_err("Buffer has no data."));
+        }
+
+        Ok(Self(buffer))
+    }
+}
+
+impl AsRef<[u8]> for PyU8Buffer {
+    fn as_ref(&self) -> &[u8] {
+        let len = self.0.item_count();
+        let data = self.0.buf_ptr() as *const u8;
+        unsafe { std::slice::from_raw_parts(data, len) }
+    }
+}
+
+struct Pyf64RTreeRef {
+    buffer: PyU8Buffer,
+    metadata: TreeMetadata<f64>,
+}
+
+impl Pyf64RTreeRef {
+    fn try_new(buffer: PyU8Buffer) -> PyResult<Self> {
+        let metadata = TreeMetadata::try_new(buffer.as_ref()).unwrap();
+        Ok(Self { buffer, metadata })
+    }
+}
+
+impl AsRef<[u8]> for Pyf64RTreeRef {
+    fn as_ref(&self) -> &[u8] {
+        self.buffer.as_ref()
+    }
+}
+
+impl RTreeIndex<f64> for Pyf64RTreeRef {
+    fn boxes(&self) -> &[f64] {
+        self.metadata.boxes_slice(self.as_ref())
+    }
+
+    fn indices(&self) -> Indices {
+        self.metadata.indices_slice(self.as_ref())
+    }
+
+    fn level_bounds(&self) -> &[usize] {
+        self.metadata.level_bounds()
+    }
+
+    fn node_size(&self) -> usize {
+        self.metadata.node_size()
+    }
+
+    fn num_items(&self) -> usize {
+        self.metadata.num_items()
+    }
+
+    fn num_nodes(&self) -> usize {
+        self.metadata.num_nodes()
+    }
+}
+
+struct Pyf32RTreeRef {
+    buffer: PyU8Buffer,
+    metadata: TreeMetadata<f32>,
+}
+
+impl Pyf32RTreeRef {
+    fn try_new(buffer: PyU8Buffer) -> PyResult<Self> {
+        let metadata = TreeMetadata::try_new(buffer.as_ref()).unwrap();
+        Ok(Self { buffer, metadata })
+    }
+}
+
+impl AsRef<[u8]> for Pyf32RTreeRef {
+    fn as_ref(&self) -> &[u8] {
+        self.buffer.as_ref()
+    }
+}
+
+impl RTreeIndex<f32> for Pyf32RTreeRef {
+    fn boxes(&self) -> &[f32] {
+        self.metadata.boxes_slice(self.as_ref())
+    }
+
+    fn indices(&self) -> Indices {
+        self.metadata.indices_slice(self.as_ref())
+    }
+
+    fn level_bounds(&self) -> &[usize] {
+        self.metadata.level_bounds()
+    }
+
+    fn node_size(&self) -> usize {
+        self.metadata.node_size()
+    }
+
+    fn num_items(&self) -> usize {
+        self.metadata.num_items()
+    }
+
+    fn num_nodes(&self) -> usize {
+        self.metadata.num_nodes()
+    }
+}
+
+pub(crate) enum PyRTreeRef {
+    Float32(Pyf32RTreeRef),
+    Float64(Pyf64RTreeRef),
+}
```

Comment:
I think we'll essentially be able to replace `RTreeInner` with this. Instead of having `RTree.from_interleaved` as a class method, we can make it a function instead

Or...